modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/H2IndexCostedBase.java

```diff
@@ -641,22 +664,558 @@ else if (needsToReadFromScanIndex)
+                        Value curVal = cond.getCurrentValue(ses);
+                        if (null == minValue || (curVal != null || filter.getTable().compareValues(minValue, curVal) < 0))
+                            minValue = curVal;
+                    }
+                }
+            }
+            return minValue;
+        }
+
+        /**
+         * Check if specified filter compare specified column to not null.
+         *
+         * @param ses    session to resolv values
+         * @param column column to check
+         * @param filter filter
+         * @return {@code true} if column value should be null, {@code falce} otherwise (or if it not sure)
+         */
+        private boolean isNotNullFilter(Session ses, Column column, TableFilter filter) {
+            // TODO: check not null expression (TableFilter contains only fullCondition without getter to check it)
+            return false;
+        }
+
+        /**
+         * Check if specified filter compare specified column to null.
+         *
+         * @param ses    session to resolv values
+         * @param column column to check
+         * @param filter filter
+         * @return {@code true} if column value should be null, {@code falce} otherwise (or if it not sure)
+         */
+        private boolean isNullFilter(Session ses, Column column, TableFilter filter) {
+            if (filter == null)
+                return false;
+            for (IndexCondition cond : filter.getIndexConditions()) {
+                if (column.equals(cond.getColumn()))
+                    continue;
+
+                if (isByteFlag(cond.getCompareType(), Comparison.SPATIAL_INTERSECTS) && cond.isEvaluatable()) {
+                    Expression expr = cond.getExpression();
+
+                    if (expr != null && expr.isConstant()) {
+                        Value curVal = cond.getCurrentValue(ses);
+
+                        if (curVal != null && curVal.getValueType() == Value.NULL)
+                            return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Test if value contains all masks bits.
+         *
+         * @param value value to test.
+         * @param mask  mask to test by.
+         * @return {@code true} if value contains all necessary bits, {@code false} otherwise.
+         */
+        private boolean isByteFlag(int value, int mask) {
+            return (value & mask) == mask;
+        }
+
+        /**
+         * Estimate percent of selected rows by specified min/max conditions (of total rows, with nulls).
+         *
+         * @param colStat column statistics to use, if exists.
+         * @param min     lower border.
+         * @param max     higher border.
+         * @return percent of rows, selected with specified conditions (0-100)
+         */
+        private int estimatePercent(ColumnStatistics colStat, Value min, Value max) {
```

Comment:
Consider the following case: you try to query some values with condition `col1 in range 0..100`, but stats claims that all values belongs to range 200..300 and there is no nulls. It seems in this particular case estimatePercent returns 100, but actually no rows should be returned, so correct estimation is 0